core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala

```diff
@@ -537,7 +545,16 @@ class TimeSeriesShard(val dataset: Dataset,
 
   def numRowsIngested: Long = ingested
 
-  def numActivePartitions: Int = partSet.size
+  def numActivePartitions: Int = {
+    while (true) {
+      val stamp = partSetLock.tryOptimisticRead()
+      val size = partSet.size
+      if (partSetLock.validate(stamp)) {
+        return size
```

Comment:
This looks like a possibility for continuous looping. Need much caution since it is called from the ingestion path for update of gauges.

Reply:
Yup.

Reply:
Not expected, but I changed the code and added a comment.